### ISSUE TITLE
wishlist: O(N^3) solution of weighted assignment problem

### ISSUE THREAD

**Issue by argriffing** [scipy/scipy#4482]
SciPy has a maximum bipartite matching implementation in csgraph, but it is unweighted.

See also http://en.wikipedia.org/wiki/Hungarian_algorithm.
And a thread on the mailing list: http://mail.scipy.org/pipermail/scipy-user/2013-March/034375.html

**Comment by larsmans** [comment]
There's an implementation of the Hungarian algorithm in scikit-learn, is that what you're looking for?

https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/linear_assignment_.py

**Comment by argriffing** [comment]
> is that what you're looking for?

@larsmans Yes! I'd seen the original source of that code but I hadn't seen that it had made its way into scikit-learn.  My motivation is that I was curious to use it to match mixture components found by EM to the correct mixture components, for testing the EM acceleration PR by @pv.  But I guess it is overkill, and I can just make an assignment by sorting on one of the axes (e.g. the mixture probability).  I'm tempted to suggest that it could have been useful in https://github.com/scipy/scipy/pull/3717 but I don't know enough about the zeros/poles assignment problem.

**Comment by tacaswell** [comment]
attn @danielballan @nkeim

This smells very similar (or at least a more general version of) to the problem we solve in the linking part of https://github.com/soft-matter/trackpy (which is aimed at linking large sparse graphs).

**Comment by jnothman** [comment]
At https://github.com/scikit-learn/scikit-learn/issues/5103 we have proposed scipy adopts scikit-learn's implementation, as it seems users are going to scikit-learn for this and it's a bit of an unusual component for scikit-learn to maintain (i.e. it's only peripheral to any learning algorithm). Would someone like to make a PR of it?

As noted there, I've found it beneficial with sparse similarity matrices to run this algorithm over each component (densified) produced by `csgraph.connected_components` so this would be one approach to supporting sparse matrix inputs.
